plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+        wf::output_t *output;
+        std::unique_ptr<wf::keyboard_interaction_t> interaction;
+        wf::wl_listener_wrapper lock_surface_destroy;
+    };
+
+    class backup_node_t : public simple_text_node_t
+    {
+      public:
+        backup_node_t(wf::output_t *output) : simple_text_node_t()
+        {
+            set_position({0, 0});
+            // TODO: it seems better to create the node and add it to the back of the scenegraph so
+            // it will be displayed if the client crashes and its surface is destroyed.
+            // Unfortunately this causes the surface to briefly appear before the lock screen.
+            // So make the background completely transparent instead, and then add the text
+            // when the client suraface is destroyed.
+            wf::cairo_text_t::params params(
+                1280 /* font_size */,
+                wf::color_t{0.1, 0.1, 0.1, 0} /* bg_color */,
+                wf::color_t{0.9, 0.9, 0.9, 1} /* fg_color */);
+            params.rounded_rect = false;
+            set_text_params(params);
+            set_size(output->get_screen_size());
+        }
+
+        void display()
+        {
+            wf::cairo_text_t::params params(
+                1280 /* font_size */,
+                wf::color_t{0, 0, 0, 1} /* bg_color */,
+                wf::color_t{0.9, 0.9, 0.9, 1} /* fg_color */);
+            set_text_params(params);
+            // TODO: make the text smaller and display a useful message instead of a big explosion.
+            set_text("💥");
+        }
+    };
+
+    struct output_state
+    {
+        output_state(wf::output_t *output) : output(output)
+        {}
+
+        wf::output_t *output;
+        std::shared_ptr<lock_surface_node_t> surface;
+        std::shared_ptr<backup_node_t> backup_surface;
+    };
+
+    class wayfire_session_lock
+    {
+      public:
+        wayfire_session_lock(wf_session_lock_plugin *plugin, wlr_session_lock_v1 *lock) :
+            plugin(plugin), lock(lock)
+        {
+            auto& ol = wf::get_core().output_layout;
+            output_added.set_callback([this] (wf::output_added_signal *ev)
+            {
+                handle_output_added(ev->output);
+            });
+            ol->connect(&output_added);
+
+            output_removed.set_callback([this] (wf::output_removed_signal *ev)
+            {
+                handle_output_removed(ev->output);
+            });
+            ol->connect(&output_removed);
+
+            for (auto output : ol->get_outputs())
+            {
+                handle_output_added(output);
+            }
+
+            new_surface.set_callback([this] (void *data)
+            {
+                wlr_session_lock_surface_v1 *lock_surface = (wlr_session_lock_surface_v1*)data;
+                wlr_output *wo = lock_surface->output;
+                wlr_session_lock_surface_v1_configure(lock_surface,
+                    wo->width / wo->scale,
```

Comment:
I don't think this takes output transform into account. You can however find the `wf::output_t` associated with the wlr output and get its geometry - the geometry will be scaled and rotated correctly.